Sources/GRPC/Docs.docc/Proposals/0001-stub-api.md

```diff
@@ -0,0 +1,1135 @@
+    decoder: some MessageDecoder<Echo.Response>,
+    _ body: @Sendable @escaping (ClientResponse.Single<EchoResponse>) async throws -> R
+  ) async rethrows -> R
+
+  func collect<R: Sendable>(
+    request: ClientRequest.Stream<EchoRequest>,
+    encoder: some MessageEncoder<Echo.Request>,
+    decoder: some MessageDecoder<Echo.Response>,
+    _ body: @Sendable @escaping (ClientResponse.Single<EchoResponse>) async throws -> R
+  ) async rethrows -> R
+
+  func expand<R: Sendable>(
+    request: ClientRequest.Single<EchoRequest>,
+    encoder: some MessageEncoder<Echo.Request>,
+    decoder: some MessageDecoder<Echo.Response>,
+    _ body: @Sendable @escaping (ClientResponse.Stream<EchoResponse>) async throws -> R
+  ) async rethrows -> R
+
+  func update<R: Sendable>(
+    request: ClientRequest.Stream<EchoRequest>,
+    encoder: some MessageEncoder<Echo.Request>,
+    decoder: some MessageDecoder<Echo.Response>,
+    _ body: @Sendable @escaping (ClientResponse.Stream<EchoResponse>) async throws -> R
+  ) async rethrows -> R
+}
+
+extension EchoClientProtocol {
+  public func get<R: Sendable>(
+    request: ClientRequest.Single<EchoRequest>,
+    _ body: @Sendable @escaping (ClientResponse.Single<EchoResponse>) async throws -> R
+  ) async rethrows -> R {
+    // Implementation elided. Calls corresponding function on
+    // `EchoClientProtocol` specifying the encoder and decoder.
+  }
+
+  func collect<R: Sendable>(
+    request: ClientRequest.Stream<EchoRequest>,
+    _ body: @Sendable @escaping (ClientResponse.Single<EchoResponse>) async throws -> R
+  ) async rethrows -> R {
+    // Implementation elided. Calls corresponding function on
+    // `EchoClientProtocol` specifying the encoder and decoder.
+  }
+
+  func expand<R: Sendable>(
+    request: ClientRequest.Single<EchoRequest>,
+    _ body: @Sendable @escaping (ClientResponse.Stream<EchoResponse>) async throws -> R
+  ) async rethrows -> R {
+    // Implementation elided. Calls corresponding function on
+    // `EchoClientProtocol` specifying the encoder and decoder.
+  }
+
+  func update<R: Sendable>(
+    request: ClientRequest.Stream<EchoRequest>,
+    _ body: @Sendable @escaping (ClientResponse.Stream<EchoResponse>) async throws -> R
+  ) async rethrows -> R {
+    // Implementation elided. Calls corresponding function on
+    // `EchoClientProtocol` specifying the encoder and decoder.
+  }
+}
+
+// Note: a concerete client implementation would also be generated. The details
+// aren't interesting here.
+```
+
+#### Example: Echo client usage
+
+An example of using the Echo client follows. Some functions highlight the
+"sugared" API built on top of the more verbose lower-level API:
+
+```swift
+func get(echo: some EchoClientProtocol) async throws {
+  // Make the request:
+  let request = ClientRequest.Single(message: Echo.Request(text: "foo"))
+
+  // Execute the RPC (most verbose API):
+  await echo.get(request: request) { response in
```

Comment:
Hi, 

Why is the API modeled in such a way that the consumer would need to pass a closure in order to use the `response`?
Wouldn't it be more straightforward to just do 
```
let response = echo.get(request: request) // response: ClientResponse.Single<EchoResponse>
switch response.result {
  ...
}
```

Reply:
This is different to a "traditional" completion handler: here the `echo.get` won't return until the handler has run which means we _know_ that after `echo.get` returns, all resources associated with the call have been freed. 

For unary RPCs this isn't actually an issue because once there's a single response the RPC is done, you can just return the `response` from the closure. We may actually default the implementation of the closure to do this in generated code.

For RPCs which stream requests and responses we can't do this without resorting to unstructured concurrency which we absolutely want to avoid: cancellation is not propagated to unstructured tasks which makes the underlying code and code which uses these APIs difficult to reason about.

Reply:
Even for the stream requests, the `body` closure will be executed only once. 
In my mind, the same syntax should work for stream requests as well, i.e
```
let response = try await echo.expand(request: request)
for try await message in response.messages {
  ...
}
```

Reply:
The the important part here is this:

> For RPCs which stream requests and responses we can't do this **without resorting to unstructured concurrency** which we absolutely want to avoid: cancellation is not propagated to unstructured tasks which makes the underlying code and code which uses these APIs difficult to reason about.

The `echo.expand` call is hiding a `TaskGroup` which runs a number of tasks required to actually execute the RPC. The response handler is running as one of those tasks.

Reply:
I see. Now that makes sense. 
Thanks for your explanation!